core/src/main/java/hivemall/knn/similarity/DIMSUMMapperUDTF.java

```diff
@@ -0,0 +1,349 @@
+        fieldNames.add("j");
+        fieldNames.add("k");
+        fieldNames.add("b_jk");
+
+        ArrayList<ObjectInspector> fieldOIs = new ArrayList<ObjectInspector>();
+        if (parseFeatureAsInt) {
+            fieldOIs.add(PrimitiveObjectInspectorFactory.writableIntObjectInspector);
+            fieldOIs.add(PrimitiveObjectInspectorFactory.writableIntObjectInspector);
+        } else {
+            fieldOIs.add(PrimitiveObjectInspectorFactory.writableStringObjectInspector);
+            fieldOIs.add(PrimitiveObjectInspectorFactory.writableStringObjectInspector);
+        }
+        fieldOIs.add(PrimitiveObjectInspectorFactory.writableFloatObjectInspector);
+
+        return ObjectInspectorFactory.getStandardStructObjectInspector(fieldNames, fieldOIs);
+    }
+
+    @Override
+    public void process(Object[] args) throws HiveException {
+        Feature[] row = parseFeatures(args[0]);
+        if (row == null) {
+            return;
+        }
+        this.probes = row;
+
+        // since the 2nd argument (column norms) is consistent,
+        // column-related values, `colNorms` and `colProbs`, should be cached
+        if (colNorms == null || colProbs == null) {
+            final int numCols = colNormsOI.getMapSize(args[1]);
+
+            if (sqrtGamma == Double.POSITIVE_INFINITY) { // set default value to `gamma` based on `threshold`
+                if (threshold > 0.d) { // if `threshold` = 0, `gamma` is INFINITY i.e. always accept <j, k> pairs
+                    this.sqrtGamma = Math.sqrt(10 * Math.log(numCols) / threshold);
+                }
+            }
+
+            this.colNorms = new HashMap<Object, Double>(numCols);
+            this.colProbs = new HashMap<Object, Double>(numCols);
+            final Map<Object, Object> m = (Map<Object, Object>) colNormsOI.getMap(args[1]);
+            for (Map.Entry<Object, Object> e : m.entrySet()) {
+                Object j = e.getKey();
+                if (parseFeatureAsInt) {
+                    j = HiveUtils.asJavaInt(j);
+                } else {
+                    j = j.toString();
+                }
+
+                double norm = HiveUtils.asJavaDouble(e.getValue());
+                if (norm == 0.d) { // avoid zero-division
+                    norm = 1.d;
+                }
+
+                colNorms.put(j, norm);
+
+                double p = Math.min(1.d, sqrtGamma / norm);
+                colProbs.put(j, p);
+            }
+        }
+
+        if (parseFeatureAsInt) {
+            forwardAsIntFeature(row);
+        } else {
+            forwardAsStringFeature(row);
+        }
+    }
+
+    private void forwardAsIntFeature(@Nonnull Feature[] row) throws HiveException {
+        final int length = row.length;
+
+        Feature[] rowScaled = new Feature[length];
+        for (int i = 0; i < length; i++) {
+            int j = row[i].getFeatureIndex();
+
+            double norm = colNorms.get(j).doubleValue();
```

Comment:
NPE could be happen in `colNorms.get(j)`

`PrimitiveUtils.doubleValue(colNorms.get(j), 0.d);` to be null safe.